cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java

```diff
@@ -0,0 +1,616 @@
+                                p.complete(new KafkaClusterRebalanceStatusBuilder()
+                                        .withSessionId(null)
+                                        .addNewCondition().withNewType(State.Stopped.toString()).endCondition().build());
+                            });
+                        } else {
+                            log.info("Getting Cruise Control rebalance user task status");
+                            apiClient.getUserTaskStatus(host, 9090 /* TODO: need public constant from CruiseControl model */, sessionId).setHandler(userTaskResult -> {
+                                if (userTaskResult.succeeded()) {
+                                    CruiseControlResponse response = userTaskResult.result();
+                                    JsonObject taskStatusJson = response.getJson();
+                                    String taskStatus = taskStatusJson.getString("Status");
+                                    if (taskStatus.equals("completed")) {
+                                        vertx.cancelTimer(t);
+                                        p.complete(new KafkaClusterRebalanceStatusBuilder()
+                                                .withSessionId(null)
+                                                .withOptimizationResult(taskStatusJson.getJsonObject("rebalance").getMap())
+                                                .addNewCondition().withType(State.Ready.toString()).endCondition().build());
+                                    } else if (taskStatus.equals("aborted")) {
+                                        log.error("Unexpected state {}", taskStatus);
+                                        vertx.cancelTimer(t);
+                                        p.fail("Unexpected state " + taskStatus);
+                                    }
+                                } else {
+                                    log.error("Cruise Control getting rebalance task status failed", userTaskResult.cause());
+                                    vertx.cancelTimer(t);
+                                    p.fail(userTaskResult.cause());
+                                }
+                            });
+                        }
+                    } else {
+                        log.error("Cruise Control getting rebalance resource failed", getResult.cause());
+                        vertx.cancelTimer(t);
+                        p.fail(getResult.cause());
+                    }
+                });
+
+            });
+        } else {
+            p.complete(clusterRebalance.getStatus());
+        }
+        return p.future();
+    }
+
+    /**
+     * This method handles the transition from {@code Stopped} state.
+     * If the user set strimzi.io/rebalance=restart annotation, it calls the Cruise Control API for requesting a new rebalance proposal.
```

Comment:
Remind me why we do this here, rather than just transitioning back to the new state (which would avoid duplicating the logic for interacting with CC). It might be necessary to rename `New` to something like `RequiresProposal`.

Reply:
It's long time I was thinking about that. I will push a change with just one state instead of both `New` and `Stopped`

Reply:
The problem I see is about `strimzi.io/rebalance` annotation when the resource is created vs when the rebalancing/proposal is stopped.
If the KCR resource is created with the annotation whatever its value is (so ignoring the annotation itself), then we should move to `PendingProposal` asking CC for a rebalance proposal.
At the same time if we are in `RequiresProposal` after stopping a rebalance/proposal we should move to `PendingProposal` only with the `restart` annotation.
So we are in a situation where with a single state we should distinguish between:
- resource just created? no matter if it has the annotation, just start a proposal request
- resource already exists and we are here after a stop operation? a proposal request should be made only with `restart` annotation value.
One way I see is to put the `Action` from the watcher down the call chain until the state machine. Wdyt?

Reply:
Thinking more, the same call is made during a normal reconciliation loop in the `createOrUpdate` where we can't understand if the resource was created or updated actually :-(

Reply:
So there are two ways to get to the `New` state:

1. When the resource is created initially (in which case the annotation could have any value). Perhaps we should ensure it doesn't already have "approve" though?
2. When the resource is transitioning from the stopped state. In this case the annotation will have the value "restart". 

So perhaps if the annotation has _any_ value when it enters the `New` state (i.e. in `onNew()`) remove the annotation, causing a self transition (`New` → `New`), and only proceed with current logic for `New` when the annotation is not present?

Reply:
> When the resource is created initially (in which case the annotation could have any value). Perhaps we should ensure it doesn't already have "approve" though?

I don't think having "approve" matters because this value makes sense only when you are in `ProposalReady`. I think that in `New` state, whatever is the annotation value (if any), it should just start rebalance proposal moving to `PendingProposal`

> So perhaps if the annotation has any value when it enters the New state (i.e. in onNew()) remove the annotation, causing a self transition (New → New), and only proceed with current logic for New when the annotation is not present?

I didn't get it :-(

Resource ADDED + ANY annotation --> PendingProposal
Resrource MODIFIED + ONLY annotation=restart --> PendingProposal

Without information about resource created or modified how you can say that you are actually in a stopped state and not in a new state so that you can handle ANY annotation or just ONLY "restart"?

Reply:
> I don't think having "approve" matters because this value makes sense only when you are in ProposalReady.

Yes, but if you don't remove the approve annotation then you're effectively pre-approving the rebalance, right?

I was thinking:

Resource ADDED + ANY annotation --> `New`
`Stopped` with restart annotation --> `New`
`New` with any non-null annotation --> `New` with a null annotation
`New` with null annotation  --> `PendingProposal`

Reply:
> Yes, but if you don't remove the approve annotation then you're effectively pre-approving the rebalance, right?

Well because the state will be moved to `ProposalPending`, the annotation will be removed anyway so it will be not there anymore when in `ProposalReady`. When a state change happens, the annotation is always removed.

Reply:
As for offline chat, let's stay with these two statuses, just trying to factor out a common method because `New` and `Stopped` have more in common.